File: Codechef/NOV19B/PHCUL.cpp
#include <iostream>
#include <cmath>
#include <iomanip>
#include <vector>

using namespace std;

const double dist(pair<int, int>& u, pair<int, int>& v) {
    int x1 = u.first, y1 = u.second;
    int x2 = v.first, y2 = v.second;
    long long int t1 = (1LL * x2 - x1) * (x2 - x1);
    long long int t2 = (1LL * y2 - y1) * (y2 - y1);
    return sqrt(t1 + t2);
}

int main() {
    int t;
    cin >> t;
    for (int test = 0; test < t; test++) {
        pair<int, int> x;
        cin >> x.first >> x.second;
        int n, m, k;
        cin >> n >> m >> k;
        vector<pair<int, int>> N(n), M(m), K(k);
        vector<int> min_n2k(n), min_m2k(m);
        for (auto &e: N) {
            cin >> e.first >> e.second;
        }
        for (auto &e: M) {
            cin >> e.first >> e.second;
        }
        for (auto &e: K) {
            cin >> e.first >> e.second;
        }
        for (int i = 0; i < n; i++) {
            double min_dist = dist(N[i], K[0]);
            for (int j = 0; j < k; j++) {
                double curr_dist = dist(N[i], K[j]);
                if (curr_dist < min_dist) {
                    min_dist = curr_dist;
                    min_n2k[i] = j;
                }
            }
        }
        for (int i = 0; i < m; i++) {
            double min_dist = dist(M[i], K[0]);
            for (int j = 0; j < k; j++) {
                double curr_dist = dist(M[i], K[j]);
                if (curr_dist < min_dist) {
                    min_dist = curr_dist;
                    min_m2k[i] = j;
                }
            }
        }
        double min_distance = dist(N[0], x) + dist(M[0], N[0]) + dist(K[0], M[0]);
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                min_distance = min(min_distance, dist(N[i], x) + dist(M[j], N[i]) + dist(K[min_m2k[j]], M[j]));
                min_distance = min(min_distance, dist(M[j], x) + dist(N[i], M[j]) + dist(K[min_n2k[i]], N[i]));
            }
        }
        cout << setprecision(10) << fixed << min_distance << "\n";
    }
    return 0;
}

File: CodeChef/NOV19B/PHCUL.py
import itertools
import math

def dist(u, v):
    x1, y1 = u
    x2, y2 = v
    return math.sqrt((x2 - x1) ** 2 + (y2 - y1) ** 2)

def grouper(iterable, n=2, fillvalue=None):
    args = [iter(iterable)] * n
    return itertools.zip_longest(*args, fillvalue=fillvalue)

if __name__ == '__main__':
    t = int(input())
    for test in range(t):
        x = tuple(map(int, input().split()))
        n, m, k = map(int, input().split())
        N = list(grouper(list(map(int, input().split()))))
        M = list(grouper(list(map(int, input().split()))))
        K = list(grouper(list(map(int, input().split()))))

        min_n2k = {}
        for i in range(n):
            min_n2k[i] = K.index(min(K, key=lambda k: dist(k, N[i])))
        min_m2k = {}
        for j in range(m):
            min_m2k[j] = K.index(min(K, key=lambda k: dist(k, M[j])))

        min_distance = dist(N[0], x) + dist(M[0], N[0]) + dist(K[0], M[0])
        for i in range(n):
            for j in range(m):
                min_distance = min(min_distance, dist(N[i], x) + dist(M[j], N[i]) + dist(K[min_m2k[j]], M[j]))
                min_distance = min(min_distance, dist(M[j], x) + dist(N[i], M[j]) + dist(K[min_n2k[i]], N[i]))
        print(min_distance)
            
